### PR TITLE
DR-1560 Correct DRS Error Codes for Failed Auth

### PR DESCRIPTION
**Pull request by tlangs**
Instead of a 500 error being returned for authentication/authorization errors, return the right 4xx error.

### CODE REVIEW

File: src/test/java/bio/terra/integration/DataRepoClient.java [hunk 1]
```diff
@@ -186,6 +186,15 @@ private String getLocationHeader(DataRepoResponse<JobModel> jobModelResponse) {
         return makeDrsRequest(path, HttpMethod.GET, entity, responseClass);
     }
 
+    public ResponseEntity<String> madeUnauthenticatedDrsRequest(String path,
```
Comment (pshapiro4broad) [inline src/test/java/bio/terra/integration/DataRepoClient.java]: I think you want present test not past here.
```suggestion
    public ResponseEntity<String> makeUnauthenticatedDrsRequest(String path,
```

File: src/main/java/bio/terra/service/iam/AuthenticatedUserRequest.java [hunk 2]
```diff
@@ -55,7 +55,7 @@ public AuthenticatedUserRequest token(Optional<String> token) {
     @JsonIgnore
     public String getRequiredToken() {
         if (!token.isPresent()) {
-            throw new ApiException("Token required");
+            throw new IamUnauthorizedException("An OAuth token is required.");
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/iam/AuthenticatedUserRequest.java:58]: Maybe I don't fully get the flow of operation but if we're running behind an OIDC proxy and we get a request without a token, isn't that a 500? How does a request arrive without a token that isn't an internal server error?

File: src/main/java/bio/terra/service/iam/sam/SamRetry.java [hunk 3]
```diff
@@ -65,6 +66,8 @@ static void retry(ConfigurationService configService, SamVoidFunction function)
                 return function.apply();
             } catch (ApiException ex) {
                 handleApiException(ex);
+            } catch (IamUnauthorizedException ex) {
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/iam/sam/SamRetry.java:69]: This seems a bit odd to me. The exception is already being thrown in the next catch handler. The difference is the top level exception type; the wrapped exception is still present and could be used to generate the correct response. Or is there no way to change the response behavior based on a wrapped exception?

File: src/main/java/bio/terra/service/iam/AuthenticatedUserRequest.java (hunk 2 above)
Comment (tlangs) [reply src/main/java/bio/terra/service/iam/AuthenticatedUserRequest.java:58]: Well, in that case, the error isn't really _internal_, and from TDR's perpective, its an unauthorized request. The exception handling is a little weird in this area of the code. The two `ApiExceptions` in this and `SamRetry` are in different packages.

File: src/main/java/bio/terra/service/iam/sam/SamRetry.java (hunk 3 above)
Comment (tlangs) [reply src/main/java/bio/terra/service/iam/sam/SamRetry.java:69]: As stated above, the `ApiException` being caught here is different from the `ApiException` being thrown by `AuthenticatedUserRequest::getRequiredToken`

Comment (pshapiro4broad) [reply src/main/java/bio/terra/service/iam/sam/SamRetry.java:69]: OK, that is confusing

**Review by pshapiro4broad — APPROVED** [review]

**Review by fboulnois — APPROVED** [review]
lgtm, I like that every `400` condition is tested.

**Review by tcjiang9 — APPROVED** [review]

**Pull request merged**